### PR TITLE
fcitx5: improve flexibility

### PR DESCRIPTION
**Pull request by name-snrl**
Add options to disable the systemd user service and override `home.sessionVariables`.

The systemd user service is often unnecessary when the desktop environment supports XDG Autostart. Overriding `home.sessionVariables` allows users to define the environment variables appropriate for their desktop environment.

cc @Kranzes @khaneliman

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -A dev --run treefmt`.

- [x] Code tested through `nix build .#test-all`
      or a targeted `nix run .#tests -- <pattern>`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [ ] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

### CODE REVIEW

**Comment by teto** [comment]
you can already disable the service via `systemd.user.services."fcitx5".enable = lib.mkForce true;` no need to add yet another option.

same comment for `overrideSessionVariables`

**Comment by name-snrl** [comment]
> you can already disable the service via systemd.user.services.\"fcitx5\".enable = lib.mkForce true; no need to add yet another option.

No, I can't, there's no such option in home manager.

> same comment for `overrideSessionVariables`

The problem is that the module may change, and I can't just unset the variables defined by this module. This wouldn't be a problem if Fcitx were configured the same way for every DE, but that's not the case. Therefore, I think it's reasonable to give user the ability to define these variables themselves or at least unset them

**Comment by teto** [comment]
> No, I can't, there's no such option in home manager.

you can set  `systemd.user.services."fcitx5".Install.WantedBy = []`.

>  I can't just unset the variables defined by this module

You can with `systemd.user.services."fcitx5".Service.Environment = lib.mkForce []`.

File: modules/i18n/input-method/fcitx5.nix [hunk 1]
```diff
@@ -201,18 +230,22 @@ in
...
-      }
-      // lib.optionalAttrs (!cfg.waylandFrontend) {
-        GTK_IM_MODULE = "fcitx";
-        QT_IM_MODULE = "fcitx";
-      }
-      // lib.optionalAttrs cfg.ignoreUserConfig {
-        SKIP_FCITX_USER_PATH = "1";
-      };
+      sessionVariables =
+        if cfg.overrideSessionVariables != { } then
```
Comment (khaneliman) [inline modules/i18n/input-method/fcitx5.nix]: issue (blocking): Could this use `null` as the default sentinel instead? Then `sessionVariables = null` could select generated defaults, while an attribute set—including `{ }`—would replace them exactly. The `ignoreUserConfig` contribution should be appended afterward so a custom set does not drop `SKIP_FCITX_USER_PATH`.

**Review by khaneliman — COMMENTED** [review]
I think both use cases are reasonable. Clearing `WantedBy` is only a workaround and still leaves the generated unit, so a module-level toggle makes sense; I would expose it as `fcitx5.systemd.enable`.

`Service.Environment` is separate from `home.sessionVariables`, so a scoped variable option also has value. I left one API suggestion inline. Please add tests for the disabled service, empty/custom variables, and `ignoreUserConfig`.

File: modules/i18n/input-method/fcitx5.nix (hunk 1 above)
Comment (name-snrl) [reply modules/i18n/input-method/fcitx5.nix]: What do you think about renaming `overrideSessionVariables` to `sessionVariables` with the default value: 

```nix
{ 
  GLFW_IM_MODULE = "ibus";
  SDL_IM_MODULE = "fcitx"; 
  XMODIFIERS = "@im=fcitx"; 
} 
// lib.optionalAttrs (!cfg.waylandFrontend) { 
  GTK_IM_MODULE = "fcitx"; 
  QT_IM_MODULE = "fcitx"; 
}
```

Comment (name-snrl) [reply modules/i18n/input-method/fcitx5.nix]: As a replacement of using `null`

**Comment by name-snrl** [comment]
@khaneliman could you take a look, please? I've added tests, renamed the options and used a different approach for `sessionVariables`

File: modules/i18n/input-method/fcitx5.nix [hunk 2]
```diff
@@ -50,6 +62,39 @@ in
...
       };
 
+      sessionVariables = lib.mkOption {
+        inherit (options.home.sessionVariables) type;
+        default = {
+          GLFW_IM_MODULE = "ibus"; # IME support in kitty
+          SDL_IM_MODULE = "fcitx";
+          XMODIFIERS = "@im=fcitx";
+        }
+        // lib.optionalAttrs (!cfg.waylandFrontend) {
```
Comment (khaneliman) [inline modules/i18n/input-method/fcitx5.nix]: Would probably move this down to below since it's config based on option and not default behavior.

File: modules/i18n/input-method/fcitx5.nix [hunk 3]
```diff
@@ -50,6 +62,39 @@ in
...
+          GTK_IM_MODULE = "fcitx";
+          QT_IM_MODULE = "fcitx";
+        };
+        defaultText = lib.literalExpression ''
+          {
+            GLFW_IM_MODULE = "ibus";
+            SDL_IM_MODULE = "fcitx";
+            XMODIFIERS = "@im=fcitx";
+          }
+          // lib.optionalAttrs (!config.i18n.inputMethod.fcitx5.waylandFrontend) {
```
Comment (khaneliman) [inline modules/i18n/input-method/fcitx5.nix]: This doesn't seem like a default if we're hiding it behind another option

**Comment by name-snrl** [comment]
@khaneliman fixed

File: tests/modules/i18n/input-method/fcitx5/user-session-variables.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+  nmt.script =
+    lib.concatMapAttrsStringSep "\n" (
+      varName: _: "assertFileNotRegex home-path/etc/profile.d/hm-session-vars.sh ${varName}"
+    ) options.i18n.inputMethod.fcitx5.sessionVariables.default
+    + ''
+      assertFileRegex home-path/etc/profile.d/hm-session-vars.sh 'SKIP_FCITX_USER_PATH'
+      assertFileRegex home-path/etc/profile.d/hm-session-vars.sh 'NIXOS_OZONE_WL'
+      assertFileRegex home-path/etc/profile.d/hm-session-vars.sh 'XMODIFIERS'
+    '';
```
Comment (khaneliman) [inline tests/modules/i18n/input-method/fcitx5/user-session-variables.nix]: issue (blocking): `concatMapAttrsStringSep` emits no trailing newline, so the generated script joins the final negative assertion to the first appended command as `XMODIFIERSassertFileRegex`. NMT treats that as one non-matching negative assertion and ignores the extra arguments, which is why CI passed while the intended `XMODIFIERS` and `SKIP_FCITX_USER_PATH` checks did not run correctly. Use explicit assertions here; `options` can then be removed from the module arguments.

```suggestion
  nmt.script = ''
    assertFileNotRegex home-path/etc/profile.d/hm-session-vars.sh 'GLFW_IM_MODULE'
    assertFileNotRegex home-path/etc/profile.d/hm-session-vars.sh 'SDL_IM_MODULE'
    assertFileContains home-path/etc/profile.d/hm-session-vars.sh 'export SKIP_FCITX_USER_PATH="1"'
    assertFileContains home-path/etc/profile.d/hm-session-vars.sh 'export NIXOS_OZONE_WL="1"'
    assertFileContains home-path/etc/profile.d/hm-session-vars.sh 'export XMODIFIERS="@im=fcitx"'
  '';
```

**Review by khaneliman — COMMENTED** [review]
One test issue remains.

**Comment by name-snrl** [comment]
@khaneliman done. Thanks for the review. It was first time I writing tests for home-manager, I should've checked the test script. Sorry about that.
